### ISSUE TITLE
OC 8.0: s2s does not recognise certificates from personal store (fix included)

### ISSUE THREAD

**Issue by tjaehnel** [owncloud/core#14790]
The problem appears in OC 8.0.0 as well as the current GitHub master.

**Reproduce:**
1. Import CACert root certificate (or any other that is not in the global store) into personal certificate store
2. Accept a sharing from another cloud that uses a certificate, signed by the just imported root certificate
3. Open the new shared folder
-> Nothing happens / you receive an "Unknown Error"

**Problem:**
The **Certificate Manager** returns the certificate store path **relative to the data directory** in *getCertificateBundle()*. (see: [lib/private/security/certificatemanager.php](https://github.com/owncloud/core/blob/master/lib/private/security/certificatemanager.php#L141), Lines: *137/141*)
As a result, the sharing code feeds cURL with this relative path, which it obviously cannot find (see: [apps/files_sharing/lib/external/storage.php](https://github.com/owncloud/core/blob/master/apps/files_sharing/lib/external/storage.php#L231) Lines: *229/231*).

**Solution:**
Make Certificate Manager return the absolute path to the store.
Change in file [lib/private/security/certificatemanager.php](https://github.com/owncloud/core/blob/master/lib/private/security/certificatemanager.php#L141), line *141* such that it reads:
```
$path = is_null($this->uid) ? '/files_external/' : '/' . \OC_User::getHome($this->uid) . '/files_external/';
```
**Note:**
Other parts of the OC code do avoid the certificate manager and construct the cert path store themselves. These work without issues.

**Comment by RealRancor** [comment]
Hi,

is this related to the already reported issue here: https://github.com/owncloud/core/issues/14747 ?

**Comment by tjaehnel** [comment]
Hi,

hmm, I don't think this is related. The stack trace in the mentioned issue references to *dav.php*, which does not use the certificate manager to determine the path to the store. See [here](https://github.com/owncloud/core/blob/master/lib/private/files/storage/dav.php#L59).

But it definitely relates to commit 67da1f7e5ac9fc5472b6041dcfdc81a4b8428379 .
There was the *user object* removed from certificate manager and replaced by *uid*, which generally is a good thing.
Though in the same step *getPathToCertificates()* was changed such that it now does not return the full path anymore, but *file_sharing* (and a few others) rely on this.

Toby

**Comment by Colely** [comment]
I tried S2S sharing with a folder from an owncloud 7.02 Server to a 8.02 Server, imported the self-signed certificate and made your alteration, but it still doesn't work. Is there a way to fix this?

**Comment by DeepDiver1975** [comment]
@LukasReschke certs'n'stuff - thx

**Comment by tjaehnel** [comment]
My patch did not work completely, because some parts of the code relate on absolute, some on relative paths.
But it seems that commit 5f044ebf1bc6f45acec2e79dc05185acd0405f42 fixes this bug and makes things much more clean - so my solution is obsolete anyway.

As a *temporary workaround* until this commit gets into one of the next releases, open file [lib/private/security/certificatemanager.php](https://github.com/owncloud/core/blob/master/lib/private/security/certificatemanager.php) and in function *getCertificateBundle* replace
`return $this->getPathToCertificates() . 'rootcerts.crt';`
with
`return \OC::$server->getConfig()->getSystemValue('datadirectory') . $this->getPathToCertificates() . 'rootcerts.crt';`

This did it for me.

**Comment by LukasReschke** [comment]
Yep - there is a lot of broken stuff in SSL / TLS handling in all current ownCloud stable releases. This is not the only thing that is broken but one of the fews that can be detected most easily.

This will be properly addressed with ownCloud 8.1 where I refactored everything HTTP related - see https://github.com/owncloud/core/pull/15195

I doubt that backporting this only single specific patch will make much sense… There is really a lot broken as stated at https://github.com/owncloud/core/pull/14913#issuecomment-82490848 and at other places.

**Comment by oparoz** [comment]
@LukasReschke - So no point reporting more SSL issues on 8.0? Fixes will never be backported?
